Add JIT for `hash?`, `syntax?`, `syntax-e`, `char<?` (etc).

samth opened the pull request:
This also makes `syntax_e` into `scheme_syntax_e` which is now exported by the
runtime.

There are a couple questions as FIXME in the code.

samth committed:
Add JIT for `hash?`, `syntax?`, `syntax-e`, `char<?` (etc).

This also makes `syntax_e` into `scheme_syntax_e` which is now
exported by the runtime.

mflatt commented on racket/src/racket/src/jitinline.c:
```diff
@@ -1283,6 +1284,12 @@ int scheme_generate_inlined_unary(mz_jit_state *jitter, Scheme_App2_Rec *app, in
     generate_inlined_type_test(jitter, app, SCHEME_PLATFORM_PATH_KIND, SCHEME_PLATFORM_PATH_KIND, 0, for_branch, branch_short, dest);
     return 1;
+  } else if (IS_NAMED_PRIM(rator, "hash?")) {
+    generate_inlined_type_test(jitter, app, scheme_hash_table_type, scheme_hash_tree_indirection_type, 1, for_branch, branch_short, dest);
```
I don't think this covers `scheme_bucket_table_type`, which is for weak hash tables

samth replied:
Should it also cover `scheme_bucket_type`?

Also, those are not in a sequence with `scheme_hash_table_type` et al. Should I move them in `stypes.h`, or write a more complex JIT function?

mflatt commented on racket/src/racket/src/jitinline.c:
```diff
@@ -1283,6 +1284,12 @@ int scheme_generate_inlined_unary(mz_jit_state *jitter, Scheme_App2_Rec *app, in
+    generate_inlined_type_test(jitter, app, scheme_hash_table_type, scheme_hash_tree_indirection_type, 1, for_branch, branch_short, dest);
+    return 1;
+  } else if (IS_NAMED_PRIM(rator, "syntax?")) {
+    generate_inlined_type_test(jitter, app, scheme_primitive_syntax_type, scheme_primitive_syntax_type, 0, for_branch, branch_short, dest);
```
This one should be `scheme_stx_type`.
The `syntax?` primitive function has to be marked as inlineable for these to take effect, so that's probably why you're not seeing failures.

mflatt commented on racket/src/racket/src/jitinline.c:
```diff
@@ -1600,6 +1607,36 @@ int scheme_generate_inlined_unary(mz_jit_state *jitter, Scheme_App2_Rec *app, in
       jit_movr_p(dest, JIT_R0);
 
       return 1;
+    } else if (IS_NAMED_PRIM(rator, "syntax-e")) {
```
Isn't `syntax-e` already JIT-inlined (further below)?

mflatt commented on racket/src/racket/src/jitinline.c:
```diff
@@ -2532,9 +2569,47 @@ static int generate_binary_char(mz_jit_state *jitter, Scheme_App3_Rec *app,
     jit_ldxi_i(JIT_R1, JIT_R1, (intptr_t)&SCHEME_CHAR_VAL((Scheme_Object *)0x0));
-    ref = jit_bner_i(jit_forward(), JIT_R0, JIT_R1);
+
+    /* FIXME: why are these tests all inverted? */
```
Because "(if [something] [t] [f])" turns into "if (not [something]) goto false; [t]; jump to end; false: [f]; end:"

mflatt commented on racket/src/racket/src/jitinline.c:
```diff
@@ -1600,6 +1607,36 @@ int scheme_generate_inlined_unary(mz_jit_state *jitter, Scheme_App2_Rec *app, in
+      __START_TINY_JUMPS__(1);
+      mz_patch_branch(ref);
+      (void)mz_bnei_t(reffail, JIT_R0, scheme_primitive_syntax_type, JIT_R1);
+      /* FIXME: do we have to care about scheme_delay_syntax_type ?*/
```
No, that's related to bytecode and no longer related to syntax; it `scheme_delay_syntax_type` should be renamed.

mflatt commented:
As noted in one comment, you'll need to marke the primitives as `SCHEME_PRIM_IS_UNARY_INLINED` or `SCHEME_PRIM_IS_BINARY_INLINED` for the JIT additions to take effect.

For consistency with recent conventions, `scheme_syntax_e` would be better as `scheme_checked_syntax_e`. The "checked" convention means "uses argc & argv".

samth committed:
Address review comments from @mflatt.

samth commented:
I think that addresses all the comments. However, the sandbox tests have started segfaulting for me. More investigation required as to whether that's related to this patch.
